### PR TITLE
Split `sign_justice_transaction` in two halves

### PR DESCRIPTION
**Pull request by ariard**
To avoid caller data struct storing HTLC-related information when
a revokeable output is claimed on top of a commitment/second-stage
HTLC transactions, we split `keysinterface::sign_justice_transaction`
in two new halves `keysinterfaces::sign_justice_revoked_output` and
`keysinterfaces::sign_justice_revoked_htlc`.

Further, this split offers more flexibility to signer policy as a
commitment revokeable output might be of a value far more significant
than HTLC ones.

Note to reviewers: Should we take opportunity to split further `sign_justice_revoked_output` to enable even better signer policy flexibility (e.g pre-signed transactions for the balance output, key handover for HTLC output w.r.t to watchtower interactions).

Required for #642

### CODE REVIEW

File: lightning/src/chain/keysinterface.rs [hunk 1]
```diff
@@ -624,7 +640,29 @@ impl BaseSign for InMemorySigner {
 		Ok((sig, htlc_sigs))
 	}
 
-	fn sign_justice_transaction(&self, justice_tx: &Transaction, input: usize, amount: u64, per_commitment_key: &SecretKey, htlc: &Option<HTLCOutputInCommitment>, secp_ctx: &Secp256k1<secp256k1::All>) -> Result<Signature, ()> {
+	fn sign_justice_revoked_output(&self, justice_tx: &Transaction, input: usize, amount: u64, per_commitment_key: &SecretKey, secp_ctx: &Secp256k1<secp256k1::All>) -> Result<Signature, ()> {
+		let revocation_key = match chan_utils::derive_private_revocation_key(&secp_ctx, &per_commitment_key, &self.revocation_base_key) {
```
Comment (devrandom) [inline lightning/src/chain/keysinterface.rs:644]:
```rust
let revocation_key = chan_utils::derive_private_revocation_key(&secp_ctx, &per_commitment_key, &self.revocation_base_key)
 .map_err(|_| ())?;
```

is a bit more concise.

File: lightning/src/chain/keysinterface.rs [hunk 2]
```diff
@@ -279,8 +279,24 @@ pub trait BaseSign {
 	#[cfg(any(test,feature = "unsafe_revoked_tx_signing"))]
 	fn unsafe_sign_holder_commitment_and_htlcs(&self, commitment_tx: &HolderCommitmentTransaction, secp_ctx: &Secp256k1<secp256k1::All>) -> Result<(Signature, Vec<Signature>), ()>;
 
-	/// Create a signature for the given input in a transaction spending an HTLC or commitment
-	/// transaction output when our counterparty broadcasts an old state.
+	/// Create a signature for the given input in a transaction spending a HTLC transction output
```
Comment (devrandom) [inline lightning/src/chain/keysinterface.rs:282]: spelling "transaction"

**Comment by devrandom** [comment]
ACK

File: lightning/src/chain/keysinterface.rs (hunk 2 above)
Comment (TheBlueMatt) [reply lightning/src/chain/keysinterface.rs:282]: Also "an HTLC" not "a HTLC". English sucks.

File: lightning/src/chain/keysinterface.rs [hunk 3]
```diff
@@ -279,8 +279,24 @@ pub trait BaseSign {
...
+	/// or a commitment transaction `to_local` output when our counterparty broadcasts an old state.
+	///
+	/// A justice transaction may claim multiples outputs at the same time if timelocks are
```
Comment (TheBlueMatt) [inline lightning/src/chain/keysinterface.rs:285]: s/multiples/multiple/

File: lightning/src/chain/keysinterface.rs [hunk 4]
```diff
@@ -279,8 +279,24 @@ pub trait BaseSign {
...
+	/// similar, but only a signature for the input at index `input` should be signed for here.
+	/// It may be called multiples time for same output(s) if a fee-bump is needed with regards
```
Comment (TheBlueMatt) [inline lightning/src/chain/keysinterface.rs:287]: s/multiples time/multiple times/

**Review by TheBlueMatt — APPROVED** [review]
Just some spelling and grammar stuff, looks good.

**Comment by ariard** [comment]
Thanks for review, I think I've addressed all the grammar stuff and took code style suggestion in 8b97f1d

File: lightning/src/chain/keysinterface.rs [hunk 5]
```diff
@@ -624,17 +640,27 @@ impl BaseSign for InMemorySigner {
...
-		};
+	fn sign_justice_revoked_output(&self, justice_tx: &Transaction, input: usize, amount: u64, per_commitment_key: &SecretKey, secp_ctx: &Secp256k1<secp256k1::All>) -> Result<Signature, ()> {
+		let revocation_key = chan_utils::derive_private_revocation_key(&secp_ctx, &per_commitment_key, &self.revocation_base_key).map_err(|_| ())?;
 		let per_commitment_point = PublicKey::from_secret_key(secp_ctx, &per_commitment_key);
-		let revocation_pubkey = match chan_utils::derive_public_revocation_key(&secp_ctx, &per_commitment_point, &self.pubkeys().revocation_basepoint) {
-			Ok(revocation_pubkey) => revocation_pubkey,
-			Err(_) => return Err(())
+		let revocation_pubkey = chan_utils::derive_public_revocation_key(&secp_ctx, &per_commitment_point, &self.pubkeys().revocation_basepoint).map_err(|_| ())?;
+		let witness_script = {
+			let counterparty_delayedpubkey = match chan_utils::derive_public_key(&secp_ctx, &per_commitment_point, &self.counterparty_pubkeys().delayed_payment_basepoint) {
```
Comment (devrandom) [inline lightning/src/chain/keysinterface.rs:648]: nit: here and a couple more places for `.map_err`, but it can also be done in a future PR

**Review by devrandom — APPROVED** [review]

File: lightning/src/chain/keysinterface.rs [hunk 6]
```diff
@@ -297,7 +313,7 @@ pub trait BaseSign {
 	/// htlc holds HTLC elements (hash, timelock) if the output being spent is a HTLC output, thus
 	/// changing the format of the witness script (which is committed to in the BIP 143
 	/// signatures).
-	fn sign_justice_transaction(&self, justice_tx: &Transaction, input: usize, amount: u64, per_commitment_key: &SecretKey, htlc: &Option<HTLCOutputInCommitment>, secp_ctx: &Secp256k1<secp256k1::All>) -> Result<Signature, ()>;
+	fn sign_justice_revoked_htlc(&self, justice_tx: &Transaction, input: usize, amount: u64, per_commitment_key: &SecretKey, htlc: &HTLCOutputInCommitment, secp_ctx: &Secp256k1<secp256k1::All>) -> Result<Signature, ()>;
```
Comment (TheBlueMatt) [inline lightning/src/chain/keysinterface.rs:316]: Oops, the doc two lines up still says "if the output being spent is an HTLC output" but it always is now.

File: lightning/src/chain/keysinterface.rs (hunk 5 above)
Comment (ariard) [reply lightning/src/chain/keysinterface.rs:648]: Yes but did it for some more switches

**Review by TheBlueMatt — APPROVED** [review]

**Pull request merged**